File: experimental/mergedserialreaderandinputstage/mergedserialreaderandinputstage.js
// All Emoncms code is released under the GNU Affero General Public License.

var serialport = require("serialport");
var SerialPort = serialport.SerialPort; // localize object constructor

var userid = 2;

var mysql = require('mysql');
var connection = mysql.createConnection({host:'localhost',user:'root',password:'raspberry',database:'emoncms'});

connection.connect(function(err) {
  console.log("mysql connected");
});

var dbinputs = {};

connection.query("SELECT id,nodeid,name,processList,record FROM input WHERE `userid` = '"+userid+"'", function(err, rows) 
{
  for (z in rows)
  {
    if (rows[z].nodeid==null) rows[z].nodeid = 0;
    if (dbinputs[rows[z].nodeid]==undefined) dbinputs[rows[z].nodeid] = {};
    dbinputs[rows[z].nodeid][rows[z].name] = {'id':rows[z].id, 'processList':rows[z].processList, 'record':rows[z].record};
  }
  //console.log(dbinputs);
});

var sp = new SerialPort("/dev/ttyAMA0", { 
  parser: serialport.parsers.readline("\n") 
});

sp.on("open", function () {
  console.log('Serial Port Open');

  sp.on('data', function(data) {

    console.log('DATA RX: ' + data);

    var values = data.split(' ');
    var nodeid = parseInt(values[1]);

    if (!isNaN(nodeid)) 
    {
      var nameid = 1;
      //unixtime = time.time()

      var d = new Date();
      var time = parseInt(d.getTime()/1000.0);
      var updatetime = d.getFullYear()+'-'+(d.getMonth()+1)+'-'+d.getDate()+" "+d.getHours()+':'+d.getMinutes()+':'+d.getSeconds();

      var test = []

      for(var i=2; i<((values.length)-1); i+=2)
      {
        // Get 16-bit integer
        var value = parseInt(values[i]) + parseInt(values[i+1])*256;
        if (value>32768) value += -65536;

        if (dbinputs[nodeid]==undefined) dbinputs[nodeid] = {};

        if (dbinputs[nodeid][nameid]==undefined) {
          console.log("Create input");
          connection.query("INSERT INTO input (userid,name,nodeid) VALUES ('"+userid+"','"+nameid+"','"+nodeid+"')", function(err, rows) {});
          dbinputs[nodeid][nameid] = true;
        } else { 
          //$input->set_timevalue($dbinputs[$nodeid][$name]['id'],$time,$value);
          if (dbinputs[nodeid][nameid]['processList']) test.push({'value':value,'processList':dbinputs[nodeid][nameid]['processList']});
        }
        //console.log(nodeid+" "+nameid+" "+value+" "+dbinputs[nodeid][nameid]['processList']);
        nameid++;
      }

      for (z in test) 
      {
        var processList = test[z]['processList'];
        var value = test[z]['value'];

        // 1. For each item in the process list
        var pairs = processList.split(',');
        for (i in pairs)    			        
        {
          var inputprocess = pairs[i].split(':');	       	            // Divide into process id and arg
          var processid = parseInt(inputprocess[0]);						      // Process id
          var arg =       parseInt(inputprocess[1]);						      // Process Arg

          if (processid==2) value = value * arg;                      // scale

          if (processid==1)
          {
            feedid = arg;
            feedname = "feed_"+feedid+"";

            // a. Insert data value in feed table
            connection.query("INSERT INTO "+feedname+" (`time`,`data`) VALUES ('"+time+"','"+value+"')", function(err, rows) {});

            // b. Update feeds table
            connection.query("UPDATE feeds SET value = '"+value+"', time = '"+updatetime+"' WHERE id='"+feedid+"'", function(err, rows) {});

            //console.log("UPDATE feeds SET value = '"+value+"', time = '"+updatetime+"' WHERE id='"+feedid+"'\n");
          }
        }
      }
    }
  });  

  sp.write("ls\n", function(err, results) {
    console.log('err ' + err);
    console.log('results ' + results);
  });  
});

File: experimental/mergedserialreaderandinputstage/mergedserialreaderandinputstage.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# All Emoncms code is released under the GNU Affero General Public License.

import serial, sys, string
import time, datetime
import MySQLdb, MySQLdb.cursors

db = MySQLdb.connect(host="localhost",user="root",passwd="raspberry",db="emoncms")
cur = db.cursor(MySQLdb.cursors.DictCursor) 
cur.execute("SELECT id,nodeid,name,processList,record FROM input WHERE `userid` = '1'")
db.autocommit(True)

dbinputs = {}
for row in cur.fetchall():
  nodeid = int(row['nodeid'])
  nameid = int(row['name'])
  
  try:
    dbinputs[nodeid]
  except KeyError:
    dbinputs[nodeid] = {}

  dbinputs[nodeid][nameid] = {'id':int(row['id']), 'processList':row['processList'], 'record':row['record']}

# Set this to the serial port of your emontx and baud rate, 9600 is standard emontx baud rate
ser = serial.Serial('/dev/ttyAMA0', 9600)

while 1:

  # Read in line of readings from emontx serial
  linestr = ser.readline()

  # Remove the new line at the end
  linestr = linestr.rstrip()

  print "DATA RX:"+linestr

  # Split the line at the whitespaces
  values = linestr.split(' ')
  nodeid = int(values[1])
  nameid = 1
  unixtime = time.time()

  test = []

  for i in range(2,(len(values)-1),2):

    # Get 16-bit integer
    value = int(values[i]) + int(values[i+1])*256
    if value>32768: value -= -65536


    if dbinputs[nodeid][nameid]['processList']!=None:
      test.append({'value':value,'processList':dbinputs[nodeid][nameid]['processList']})
   
    nameid += 1

  for item in test:

    processList = item['processList']
    value = item['value']
    # 1. For each item in the process list
    pairs = processList.split(',')

    for pair in pairs:

      inputprocess = pair.split(':')
      processid = int(inputprocess[0]) # Process ID
      arg       = float(inputprocess[1]) # Process Arg

      if processid==2: value = value * arg; # scale

      if processid==1:
        feedid = int(arg)
        feedname = "feed_"+str(feedid)

        # a. Insert data value in feed table
        cur.execute("INSERT INTO "+feedname+" (time,data) VALUES(%s,%s)",(unixtime,value))

        # b. Update feeds table
        updatetime = time.strftime('%Y-%m-%d %H:%M:%S')
        cur.execute("""UPDATE feeds SET value = %s, time = %s WHERE id = %s""",(value,str(updatetime),feedid))
        print "UPDATE feeds SET value = '"+str(value)+"', time = '"+updatetime+"' WHERE id='"+str(feedid)+"'"

